src/diem/offchain/client.py

```diff
@@ -82,28 +82,36 @@ def send_request(
             raise CommandResponseError(cmd_resp)
         return cmd_resp
 
-    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> PaymentCommand:
+    def process_inbound_request(self, request_sender_address: str, request_bytes: bytes) -> CommandVariant:
         if not request_sender_address:
             raise protocol_error(ErrorCode.missing_http_header, f"missing {http_header.X_REQUEST_SENDER_ADDRESS}")
         try:
             _, public_key = self.get_base_url_and_compliance_key(request_sender_address)
         except ValueError as e:
             raise protocol_error(ErrorCode.invalid_x_request_sender_address, str(e)) from e
 
-        request = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
-        if request.command_type == CommandType.PaymentCommand:
-            payment = typing.cast(PaymentCommandObject, request.command).payment
+        command_request_object = _deserialize_jws(request_bytes, CommandRequestObject, public_key, command_error)
+
+        if command_request_object.command_type == CommandType.PaymentCommand:
+            payment = typing.cast(PaymentCommandObject, command_request_object.command).payment
             self.validate_addresses(payment, request_sender_address)
-            cmd = self.create_inbound_payment_command(request.cid, payment)
+            cmd = self.create_inbound_payment_command(command_request_object.cid, payment)
             if cmd.is_initial():
                 self.validate_dual_attestation_limit(cmd.payment.action)
             elif cmd.is_rsend():
                 self.validate_recipient_signature(cmd, public_key)
             return cmd
+        elif command_request_object.command_type == CommandType.FundPullPreApprovalCommand:
```

Comment:
use isinstance

Reply:
I don't think `isinstance` is the right decision here. The request object must have the fields specific certain command type because the command type field says so and not the other way around. The spec defines that if the type field says so, I can count on existence of all the other fields. The semantics here are important - it shows that the decision explicitly follows the spec. Relying on Python's type inference is kind of a last resort solution meaning the object doesn't hold enough semantic information, when in fact it does.